{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9110076b-0183-499e-9f14-b31cf3d70319",
   "metadata": {},
   "outputs": [],
   "source": [
    "import libsequence\n",
    "import sys\n",
    "import pandas as pd\n",
    "import math\n",
    "import argparse\n",
    "import vcf\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d70aa596-5f15-40e0-ab7d-02da557a891f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to create dictionary of polySIM summary statistics. Returns dictionary of summary stats\n",
    "def get_polySIM_stats(sd):\n",
    "    #Create polysim object\n",
    "    ps = libsequence.PolySIM(sd)\n",
    "    #Create list of methods (ie polySIM summaryStats)\n",
    "    a = [method for method in dir(ps) if callable(getattr(ps, method)) if not method.startswith('_')]\n",
    "    #Loop through methods, storing names as keys, and estimates as values in dict\n",
    "    ss_dict = {}\n",
    "    for method in a:\n",
    "        ss_dict[method] = getattr(ps, method)()\n",
    "        \n",
    "    return(ss_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6e2f6e4-865d-459b-94fa-80820915361c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to create dictionary of LD stats. Returns dictionary.\n",
    "def get_LD_stats(sd):\n",
    "    ld = libsequence.ld(sd)\n",
    "    df = pd.DataFrame(ld)\n",
    "    ss_dict = {}\n",
    "    try:\n",
    "        ss_dict['meanrsq'] = sum(df['rsq'])/len(df['rsq'])\n",
    "        ss_dict['meanD'] = sum(df['D'])/len(df['D'])\n",
    "        ss_dict['meanDprime'] = sum(df['Dprime'])/len(df['Dprime'])\n",
    "    except Exception:\n",
    "        ss_dict['meanrsq'] = np.nan\n",
    "        ss_dict['meanD'] = np.nan\n",
    "        ss_dict['meanDprime'] = np.nan\n",
    "    return(ss_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5a59158-b9c9-4868-9eb1-31004bf5c4f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define chromosome lengths\n",
    "d_lens = {\n",
    "    'NC_071442.1':216975769,\n",
    "    'NC_071443.1':202808461,\n",
    "    'NC_071444.1':189455076,\n",
    "    'NC_071445.1':173414057,\n",
    "    'NC_071446.1':161716902,\n",
    "    'NC_071447.1':159674559,\n",
    "    'NC_071448.1':156129252,\n",
    "    'NC_071449.1':126104592,\n",
    "    'NC_071450.1':132900640,\n",
    "    'NC_071451.1':136971485,\n",
    "    'NC_071452.1':128529401,\n",
    "    'NC_071453.1':123697827,\n",
    "    'NC_071454.1':117638787,\n",
    "    'NC_071455.1':112969529,\n",
    "    'NC_071456.1':98621277,\n",
    "    'NC_071457.1':98110366,\n",
    "    'NC_071458.1':74700100,\n",
    "    'NC_071459.1':47063576,\n",
    "    'NC_071460.1':50540917,\n",
    "    'NC_071461.1':44412365,\n",
    "    'NC_071462.1':50614742,\n",
    "    'NC_071463.1':49900457\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d039e5dc-5394-4e85-99ab-a86ca8e6cfa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create bed files for 10kb windows. Bedtools intersect will be used on these to calculate number of unmasked sites.\n",
    "for chrom in d_lens.keys():\n",
    "    lst = []\n",
    "    for i in range(0, d_lens[chrom]-5000, 5000):\n",
    "        lst.append([chrom, i, i+10000])\n",
    "    df = pd.DataFrame(lst, columns=['#chr','start','end'])\n",
    "    #Ensure final window ends at chromosome end coordinate\n",
    "    df['end'] = np.where(df.end <=d_lens[chrom], df.end, d_lens[chrom])\n",
    "    df.to_csv(\"/home/vivak/chimerism/marmoset/empirical/vcf/demog/\" + chrom + \"_10kb.bed\", header=True, sep='\\t',\n",
    "                                                            index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae8cb1b0-bb2b-4ee9-80f9-4f5cb711d13c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate genome-wide stats\n",
    "#Create empty list to store input for pylibseq (this will be a nested list)\n",
    "l_data = []\n",
    "for chrom in d_lens.keys():\n",
    "    #subset mask df for chromosome and bin coordinates by windows\n",
    "    mdf = pd.read_csv(\"/home/vivak/chimerism/marmoset/empirical/vcf/demog/masked_\"+chrom+\"_10kb.bed\", names=['#chr','start','end','accessible_sites'], sep='\\t')\n",
    "    #Load vcf into python\n",
    "    vcf_reader = vcf.Reader(filename=r'/home/vivak/chimerism/marmoset/empirical/vcf/demog/ALL_'+chrom+'.snps.GT.GATK_BP.DP.repeat_constrained_exons_10kb_mask.vcf.gz')\n",
    "    \n",
    "\n",
    "    #Loop through records in vcf\n",
    "    for record in vcf_reader:\n",
    "        #Create list of relative position (index 0), and genotype as string (index 1)\n",
    "        l_data.append([record.POS, ''.join([x['GT'] for x in record.samples]).replace('/','').replace('|','')])\n",
    "\n",
    "#Create libseq object from l_data\n",
    "sd = libsequence.SimData(l_data)\n",
    "\n",
    "d = {}\n",
    "d = {**d, **get_polySIM_stats(sd)}\n",
    "for stat in ['nhaps', 'numexternalmutations', 'numpoly', 'numsingletons','rm', 'thetah', 'thetal', 'thetapi', 'thetaw']:\n",
    "                    d[stat] = d[stat] / 213936105\n",
    "df = pd.DataFrame.from_dict(d, orient='index').T\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "818db741-7b0f-40cf-9c50-496e881f23eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate stats across 10kb sliding windows\n",
    "sdf = pd.DataFrame()\n",
    "stepSize = 5000\n",
    "winSize = 10000\n",
    "for chrom in d_lens.keys():\n",
    "    #subset mask df for chromosome and bin coordinates by windows\n",
    "    mdf = pd.read_csv(\"/home/vivak/chimerism/marmoset/empirical/vcf/demog/masked_\"+chrom+\"_10kb.bed\", names=['#chr','start','end','accessible_sites'], sep='\\t')\n",
    "    #Load vcf into python\n",
    "    vcf_reader = vcf.Reader(filename=r'/home/vivak/chimerism/marmoset/empirical/vcf/demog/ALL_'+chrom+'.snps.GT.GATK_BP.DP.repeat_constrained_exons_10kb_mask.vcf.gz')\n",
    "    \n",
    "    #Create empty list to store input for pylibseq (this will be a nested list)\n",
    "    l_data = []\n",
    "    positions = []\n",
    "    #Loop through records in vcf\n",
    "    for record in vcf_reader:\n",
    "        #Create list of relative position (index 0), and genotype as string (index 1)\n",
    "        l_data.append([record.POS, ''.join([x['GT'] for x in record.samples]).replace('/','').replace('|','')])\n",
    "        #positions.append([record.POS/d_lens[chrom],record.POS])\n",
    "\n",
    "    #Create libseq object from l_data\n",
    "    sd = libsequence.SimData(l_data)\n",
    "    #define sliding windows\n",
    "    wins = libsequence.Windows(sd,window_size=winSize,step_len=stepSize,starting_pos=0,ending_pos=d_lens[chrom])\n",
    "    #Create a variable for number of windows\n",
    "    num_wins = len(wins)\n",
    "    \n",
    "    #Create empty df which will be our final output format\n",
    "    df = pd.DataFrame()\n",
    "    \n",
    "    #Loop through windows\n",
    "    for k, win in enumerate(wins):\n",
    "        if((win.size()>5) & (mdf.iloc[k].accessible_sites>50)):\n",
    "            #Create empty dict to store results\n",
    "            wd = {}\n",
    "            #Add window number to dict\n",
    "            wd['chrom'] = chrom\n",
    "            wd['window'] = k\n",
    "            S = len([x for x in l_data if (x[0]>=mdf.iloc[k].start) & (x[0]<=mdf.iloc[k].end)])/mdf.iloc[k].accessible_sites\n",
    "            if(S>0):\n",
    "                wd['S'] = S\n",
    "                wd = {**wd, **get_polySIM_stats(win)}\n",
    "                for stat in ['nhaps', 'numexternalmutations', 'numpoly', 'numsingletons','rm', 'thetah', 'thetal', 'thetapi', 'thetaw']:\n",
    "                    wd[stat] = wd[stat] / mdf.iloc[k].accessible_sites\n",
    "                #For LD things are slightly more complex as libseq throws out an error if there aren't enough SNPs. We can avoid this with an if statement\n",
    "                # if len(win.pos()) >= 5: #LD stats are pairwise. If only 1 site exists, it'll show an error.\n",
    "                #     #Add all LD stats to dict\n",
    "                #     wd = {**wd, **get_LD_stats(win)}\n",
    "                # else:\n",
    "                #     #Otherwise use 'NA' or you can use np.nan\n",
    "                #     wd['meanrsq'] = np.nan\n",
    "                #     wd['meanD'] = np.nan\n",
    "                #     wd['meanDprime'] = np.nan\n",
    "                #Convert dict into temp df        \n",
    "                tdf = pd.DataFrame.from_dict(wd, orient='index').T\n",
    "                #Concatenate temp df with main output df\n",
    "                df = pd.concat([df, tdf])\n",
    "    sdf = pd.concat([sdf, df])\n",
    "    #Write stats df out to file\n",
    "    df.to_csv(\"/home/vivak/chimerism/marmoset/empirical/stats/demog/\" + chrom + \"_10kb.stats\", header=True, sep='\\t', index=False)\n",
    "sdf.to_csv(\"/home/vivak/chimerism/marmoset/empirical/stats/demog/concat_10kb.stats\", header=True, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0407a49f-6e02-45cb-a668-5f130254c742",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get means and stds of each stat for ABC analysis\n",
    "sdf = pd.read_csv(\"/home/vivak/chimerism/marmoset/empirical/stats/demog/concat_10kb.stats\", header=0, sep='\\t')\n",
    "sdf = sdf[sdf.numpoly>0]\n",
    "means = []\n",
    "stds = []\n",
    "for stat in sdf.columns[2:]:\n",
    "    means.append(sdf[stat].mean())\n",
    "    stds.append(sdf[stat].std())\n",
    "    \n",
    "df = pd.DataFrame(means+stds).T\n",
    "df.columns = [x+'_m' for x in sdf.columns[2:]] + [x+'_std' for x in sdf.columns[2:]]\n",
    "df.to_csv(\"/home/vivak/chimerism/marmoset/empirical/stats/demog/10kb_ABC.stats\", header=True, sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
